src/mlpack/methods/lsh/lsh_search_impl.hpp

```diff
@@ -339,12 +340,301 @@ void LSHSearch<SortPolicy>::BaseCase(const size_t queryIndex,
         referenceIndex, distance);
 }
 
+
+// Compare class for <double, size_t> pair, used in GetAdditionalProbingBins.
+class CompareGreater
+{
+  public:
+    bool operator()(
+        std::pair<double, size_t> p1,
+        std::pair<double, size_t> p2){
+      //only compare the double values
+      return p1.first > p2.first;
+    }
+};
+
+//Returns the score of a perturbation vector generated by perturbation set A.
+//The score of a pertubation set (vector) is the sum of scores of the
+//participating actions.
+inline double perturbationScore(
+              const std::vector<size_t> &A,
+              const arma::vec &scores)
+{
+  double score = 0.0;
+  for (size_t i = 0; i < A.size(); ++i)
+    score+=scores[A[i]];
+  return score;
+}
+
+// Inline function used by GetAdditionalProbingBins. The vector shift operation
+// replaces the largest element of a vector A with (largest element) + 1.
+inline void perturbationShift(std::vector<size_t> &A)
+{
+  size_t max_pos = 0;
+  size_t max = A[0];
+  for (size_t i = 1; i < A.size(); ++i)
+  {
+    if (A[i] > max)
+    {
+      max = A[i];
+      max_pos = i;
+    }
+  }
+  A[max_pos]++;
+}
+
+// Inline function used by GetAdditionalProbingBins. The vector expansion
+// operation adds the element [1 + (largest_element)] to a vector A, where
+// largest_element is the largest element of A.
+inline void perturbationExpand(std::vector<size_t> &A)
+{
+  size_t max = A[0];
+  for (size_t i = 1; i < A.size(); ++i)
+    if (A[i] > max)
+      max = A[i];
+  A.push_back(max+1);
+}
+
+// Return true if perturbation set A is valid. A perturbation set is invalid if
+// it contains two (or more) actions for the same dimension or dimensions that
+// are larger than the queryCode's dimensions.
+inline bool perturbationValid(
+    const std::vector<size_t> &A,
+    const size_t numProj)
+{
+  // Stack allocation and initialization to 0 (bool check[numProj] = {0}) made
+  // some compilers complain. We use new just to be safe.
+  bool *check = new bool[numProj]();
```

Comment:
Maybe use `std::vector<bool>` here?  It avoids the need for manual memory management and I think that most compilers/stdlib implementations will pack the booleans so the required amount of memory will be smaller.  I can't remember the exact details at the moment.